keras_cv/training/simclr/simclr.py

```diff
@@ -0,0 +1,175 @@
+        # optimizer.
+        self.optimizer = contrastive_optimizer
+        self.simclr_loss = SimCLRLoss(temperature)
+        self.simclr_loss_metric = keras.metrics.Mean(name="simclr_loss")
+
+        if self.include_probing:
+            self.probe_loss = keras.losses.SparseCategoricalCrossentropy(
+                from_logits=True
+            )
+            self.probe_loss_metric = keras.metrics.Mean(name="probe_loss")
+            self.probe_accuracy = keras.metrics.SparseCategoricalAccuracy(
+                name="probe_accuracy"
+            )
+
+            if not probe_optimizer:
+                raise ValueError(
+                    "`probe_optimizer` must be specified when `include_probing` is `True`."
+                )
+            self.probe_optimizer = probe_optimizer
+
+    @property
+    def metrics(self):
+        metrics = [
+            self.simclr_loss_metric,
+        ]
+        if self.include_probing:
+            metrics += [
+                self.probe_loss_metric,
+                self.probe_accuracy,
+            ]
+        return metrics
+
+    def train_step(self, data):
+        if self.include_probing:
+            if type(data) is not tuple or len(data) != 2:
+                raise ValueError(
+                    "Targets must be provided when `include_probing` is True"
+                )
+            images, labels = data
+        else:
+            if type(data) is tuple:
+                raise ValueError(
+                    "Targets must not be provided when `include_probing` is False"
+                )
+            images = data
+
+        augmented_images_1 = self.augmenter(images, training=True)
+        augmented_images_2 = self.augmenter(images, training=True)
+
+        with tf.GradientTape() as tape:
+            features_1 = self.encoder(augmented_images_1, training=True)
+            features_2 = self.encoder(augmented_images_2, training=True)
+
+            projections_1 = self.projection_top(features_1, training=True)
+            projections_2 = self.projection_top(features_2, training=True)
+
+            simclr_loss = self.simclr_loss.call(projections_1, projections_2)
+
+        gradients = tape.gradient(
+            simclr_loss,
+            self.encoder.trainable_weights + self.projection_top.trainable_weights,
+        )
+
+        self.optimizer.apply_gradients(
+            zip(
+                gradients,
+                self.encoder.trainable_weights + self.projection_top.trainable_weights,
+            )
+        )
+        self.simclr_loss_metric.update_state(simclr_loss)
+
+        if self.include_probing:
+            with tf.GradientTape() as tape:
+                features = self.encoder(images, training=False)
+                class_logits = self.probing_top(features, training=True)
+                probe_loss = self.probe_loss(labels, class_logits)
+            gradients = tape.gradient(probe_loss, self.probing_top.trainable_weights)
```

Comment:
you may want to call `tf.stop_gradient()` after  features to not. compute the full gradient set in the gradient tape.

Reply:
I think this is already effectively accomplished by using `training=False` on the encoder call, but please correct me if I'm wrong.